### PR TITLE
WT-4414 Add configuration to allow flushing modified trees independent of checkpoints when logging is disabled

### PR DESCRIPTION
**Pull request by hbokhari**

### CODE REVIEW

**Comment by hbokhari** [comment]
@agorrod, @bvpvamsikrishna: This is a first draft of changes. Feel free to comment on flag and config naming.

A python unit test is in pipeline.

File: dist/api_data.py [hunk 1]
```diff
@@ -983,6 +983,12 @@ def __ge__(self, other):
         size and the default config would extend log files in allocations of
         the maximum log file size.''',
         type='list', choices=['data', 'log']),
+    Config('file_handle_close_sync', 'true', r'''
```
Comment (sueloverso) [inline dist/api_data.py:986]: While I'm not involved in the overall discussion, my one comment here would be to shorten the API name. This would be one of the longest names otherwise. I think that `file_close_sync` is equally clear. If you make that change please rename the WT_CONN flag also.

File: dist/api_data.py [hunk 2]
```diff
@@ -983,6 +983,12 @@ def __ge__(self, other):
...
+        flush modified files to stable storage independent of checkpoints when
+        closing existing file handles and opening an exclusive file handle. If
+        set to false, certain API calls that require exclusive access to files may
+        return EBUSY to caller''',
```
Comment (agorrod) [inline dist/api_data.py:990]: Further to Sue's comment. We generally talk about tables at the API level instead of files. "control whether to flush modified tables to table storage independent of a global checkpoint when acquiring exclusive access to a table. If set to false, API calls that require exclusive access to tables will return EBUSY if there have been changes made to the table since the last global checkpoint."

File: dist/api_data.py (hunk 1 above)
Comment (hbokhari) [reply dist/api_data.py]: Fixed

File: dist/api_data.py (hunk 2 above)
Comment (hbokhari) [reply dist/api_data.py]: Fixed.

**Comment by hbokhari** [comment]
@agorrod @bvpvamsikrishna 

This one is ready for another round of review. There are two failing unit tests: 1 - `hs06` on Linux which is a know issue due to column store features, 2 - The unit test I added is failing on Windows because the underlying Python libraries are not able to parse the EBUSY exception message. I will create another ticket to track the python issue.

File: dist/api_data.py [hunk 3]
```diff
@@ -983,6 +983,13 @@ def __ge__(self, other):
         size and the default config would extend log files in allocations of
         the maximum log file size.''',
         type='list', choices=['data', 'log']),
+    Config('file_close_sync', 'true', r'''
+        control whether to flush modified tables to storage independent
+        of a global checkpoint when acquiring exclusive access to a table.
+        If set to false, API calls that require exclusive access to tables
+        will return EBUSY if there have been changes made to the table since
+        the last global checkpoint.''',
```
Comment (sueloverso) [inline dist/api_data.py:991]: The word `close` does not appear in the description at all. If this setting is not related to close, it should be removed from the configuration string name. If it is related to close specifically, it should say that simply. If I were a user, I would not know what API calls require/acquire exclusive access to a table.

Comment (hbokhari) [reply dist/api_data.py]: @sueloverso: Thanks for pointing this out. I have added why `close` in name is relevant. I have also added some text to indicate that this config will only work when logging is disabled.

File: dist/api_data.py [hunk 4]
```diff
@@ -983,6 +983,14 @@ def __ge__(self, other):
         size and the default config would extend log files in allocations of
         the maximum log file size.''',
         type='list', choices=['data', 'log']),
+    Config('file_close_sync', 'true', r'''
+        control whether to flush modified tables to storage independent
```
Comment (sueloverso) [inline dist/api_data.py:987]: Overall, I think we need something reworded because the only time this setting can be effective is if logging is disabled. I think it should start with that prominently. Maybe something like:
"when false, and logging is disabled, API calls that require exclusive access to tables will return EBUSY if there have been changes made to the table since the last global checkpoint. When true, or when logging is enabled, modified tables will be flushed to storage when closing table when required for APIs needing exclusive access."

File: src/txn/txn_ckpt.c [hunk 5]
```diff
@@ -1824,11 +1825,14 @@ __wt_checkpoint_close(WT_SESSION_IMPL *session, bool final)
     }
 
     /*
-     * Don't flush data from trees when there is a stable timestamp set: that can lead to files that
-     * are inconsistent on disk after a crash.
+     * Don't flush data from modified trees independent of system-wide checkpoint if logging is
+     * disabled, and, either there is a stable timestamp set or the connection is configured to
+     * disallow such operation. Flushing trees with logging disabled can lead to files that are
+     * inconsistent on disk after a crash.
```
Comment (sueloverso) [inline src/txn/txn_ckpt.c:1831]: We should avoid the phrase "logging is disabled" in this comment because none of the conditions are looking explicitly at logging. It is just a side effect of the `wt_btree_immediately_durable` call.

File: dist/api_data.py (hunk 4 above)
Comment (hbokhari) [reply dist/api_data.py:987]: Ah this is what I started with, somehow changed my mind and reworded it in the current form. I'll fix this.

File: src/txn/txn_ckpt.c (hunk 5 above)
Comment (hbokhari) [reply src/txn/txn_ckpt.c]: Fixed

File: src/txn/txn_ckpt.c [hunk 6]
```diff
@@ -1838,11 +1839,13 @@ __wt_checkpoint_close(WT_SESSION_IMPL *session, bool final)
...
-     * are inconsistent on disk after a crash.
+     * Don't flush data from modified trees independent of system-wide checkpoint when either there
+     * is a stable timestamp set or the connection is configured to disallow such operation.
+     * Flushing trees can lead to files that are inconsistent on disk after a crash.
      */
-    if (btree->modified && !bulk && S2C(session)->txn_global.has_stable_timestamp &&
-      !__wt_btree_immediately_durable(session))
+    if (btree->modified && !bulk && !__wt_btree_immediately_durable(session) &&
+      (S2C(session)->txn_global.has_stable_timestamp ||
+          (!F_ISSET(S2C(session), WT_CONN_FILE_CLOSE_SYNC) && !metadata && !final)))
```
Comment (sueloverso) [inline src/txn/txn_ckpt.c:1848]: In the description in `api_data.py` you say "If set to true... API calls that require exclusive access to tables will return EBUSY" yet here we are checking for 'false' to return EBUSY.

File: test/suite/test_config08.py [hunk 7]
```diff
@@ -0,0 +1,99 @@
...
+class test_config08(wttest.WiredTigerTestCase):
+
+    logging = [
+        ('log_off', dict(log='(enabled=false)')),
+        ('log_on', dict(log='(enabled=true)')),
+    ]
+    fileFlush = [
+        ('flush_off', dict(file_close_sync='false')),
+        ('flush_on', dict(file_close_sync='true')),
+    ]
```
Comment (sueloverso) [inline test/suite/test_config08.py:45]: These should just contain the true/false and add in the config setting once. Also we do not usually use camelCase. See comment below about connection too.
```
logging = [
   ('log_off', dict(logging='false')),
        ('log_on', dict(logging='true)')),
    ]
    flush = [
        ('flush_off', dict(flush='false')),
        ('flush_on', dict(flush='true')),
    ]
```

File: test/suite/test_config08.py [hunk 8]
```diff
@@ -0,0 +1,99 @@
...
+            'log=' + log + ',' + 'file_close_sync=' + file_close_sync
+        try:
+            self.conn = wiredtiger.wiredtiger_open(self.home, conn_params)
+            self.session = self.conn.open_session()
+        except wiredtiger.WiredTigerError as e:
+            print("Failed conn at '%s' with config '%s'" % (dir, conn_params))
+
+    # Create a table with logging setting matching the connection level config.
+    def test_config08(self):
+        self.ConnectionOpen(self.log, self.file_close_sync)
```
Comment (sueloverso) [inline test/suite/test_config08.py:68]: You can remove this with that function above.

File: test/suite/test_config08.py [hunk 9]
```diff
@@ -0,0 +1,99 @@
...
+
+    # This test varies the logging and file flush settings and therefore needs to set up its own
+    # connection. Override the standard methods.
+    def setUpConnectionOpen(self, dir):
+        return None
+    def setUpSessionOpen(self, conn):
+        return None
+    def ConnectionOpen(self, log, file_close_sync):
+        self.home = '.'
+        # Explicitly increase time between checkpoints
```
Comment (sueloverso) [inline test/suite/test_config08.py:57]: Several things:
1. All comments should be complete sentences, ending with a period. (There are several here that are missing that.)
2. Why do we want to have such a large time between checkpoints? Why do we need this at all? This test is adding in 3 key/value pairs so it will not take 60 seconds to run and you explicitly call `session.checkpoint` below, so I think the part about checkpoint can just be removed.
3. Why does this need a re-implementation of `ConnectionOpen`? I don't think we expect failure here. It can be simplified to setting `conn_config` as a function:
```
def conn_config(self):
            return 'log=(enabled=%s),' % self.logging + \
            'file_close_sync=%s,' % self.flush
```

File: test/suite/test_config08.py [hunk 10]
```diff
@@ -0,0 +1,99 @@
...
+        self.session.create(self.uri, table_params)
+        c = self.session.open_cursor(self.uri, None)
+        c[0] = 'ABC' * 4096
+        c[1] = 'DEF' * 4096
+        c[2] = 'GHI' * 4096
+        c.close()
+
+        # API calls that require exclusive file handles should return EBUSY if dirty tables flush
+        # and logging are disabled.
+        if self.log == '(enabled=false)' and self.file_close_sync == 'false':
```
Comment (sueloverso) [inline test/suite/test_config08.py:82]: This changes with the other change I suggest above:
if self.logging == 'false' and ...

File: test/suite/test_config08.py [hunk 11]
```diff
@@ -0,0 +1,99 @@
...
+        # Create a table with some data
+        self.session.create(self.uri, table_params)
+        c = self.session.open_cursor(self.uri, None)
+        c[0] = 'ABC' * 4096
+        c[1] = 'DEF' * 4096
+        c[2] = 'GHI' * 4096
+        c.close()
+
+        # API calls that require exclusive file handles should return EBUSY if dirty tables flush
+        # and logging are disabled.
```
Comment (sueloverso) [inline test/suite/test_config08.py:81]: The comment does not match the code. It should say "if file_close_sync is false and logging is disabled". (This also relates to my comment earlier, that the description in `api_data.py` is incorrect and should say that it returns EBUSY when set to false, not true.)

File: dist/api_data.py [hunk 12]
```diff
@@ -983,6 +983,16 @@ def __ge__(self, other):
...
+        of a global checkpoint when closing file handles to acquire exclusive
+        access to a table. If set to true, and logging disabled, API calls that
```
[the remark this answers is not in the document]
Comment (hbokhari) [reply dist/api_data.py:989]: Fixed.

File: src/txn/txn_ckpt.c (hunk 6 above)
Comment (hbokhari) [reply src/txn/txn_ckpt.c:1848]: Good catch - the documentation has wrong information. We return EBUSY when config is false.

File: test/suite/test_config08.py (hunk 7 above)
Comment (hbokhari) [reply test/suite/test_config08.py:45]: Fixed.

File: test/suite/test_config08.py (hunk 10 above)
Comment (hbokhari) [reply test/suite/test_config08.py:82]: Fixed.

File: test/suite/test_config08.py (hunk 8 above)
Comment (hbokhari) [reply test/suite/test_config08.py:68]: I copy-pasted the boiler plate code from another test because I didn't know how to change default connection config. But looks like overriding the `conn_config()` is a much cleaner approach.

File: test/suite/test_config08.py (hunk 9 above)
Comment (hbokhari) [reply test/suite/test_config08.py:57]: Fixed.

File: test/suite/test_config08.py (hunk 11 above)
Comment (hbokhari) [reply test/suite/test_config08.py:81]: Fixed.

File: test/suite/test_config08.py [hunk 13]
```diff
@@ -0,0 +1,87 @@
...
+        c[2] = 'GHI' * 4096
+        c.close()
+
+        # API calls that require exclusive file handles should return EBUSY if file_close_sync is
+        # set to false and logging is disabled.
+        if self.logging == 'false' and self.flush == 'false':
+            # WT won't allow this operation as exclsuive file handle is not possible
+            # with modified table.
+            self.assertTrue(self.raisesBusy(lambda: self.session.verify(self.uri, None)),
+                "Was expecting API call to fail with EBUSY")
```
Comment (bvpvamsikrishna) [inline test/suite/test_config08.py:74]: This msg will not be at the start, so no need to capitalize 'W'

File: src/include/wiredtiger.in [hunk 14]
```diff
@@ -2893,6 +2893,13 @@ struct __wt_connection {
  * specified to a library extension are passed to WT_CONNECTION::load_extension as the \c config
  * parameter (for example\, <code>extensions=(/path/ext.so={entry=my_entry})</code>)., a list of
  * strings; default empty.}
+ * @config{file_close_sync, control whether to flush modified tables to storage independent of a
```
[the remark this answers is not in the document]
Comment (agorrod) [reply src/include/wiredtiger.in:2896]: I asked for the switch to calling things "tables" I don't think users should know about files. OTOH: the rest of the API has been approved talking about files, so I think Vamsi is right.

[the remark this answers is not in the document]
Comment (hbokhari) [reply src/include/wiredtiger.in:2896]: Fair enough. I'll switch to word "file".

File: test/suite/test_config08.py (hunk 13 above)
Comment (hbokhari) [reply test/suite/test_config08.py]: Fixed.
